csaf/src/walker.rs

```diff
@@ -76,32 +78,72 @@ impl<S: Source> Walker<S> {
                     continue;
                 }
             }
-            log::debug!("Walking: {}", distribution.directory_url);
-            let index = self
-                .source
-                .load_index(distribution)
-                .await
-                .map_err(Error::Source)?;
-
-            let progress = self.progress.start(index.len());
-
-            for advisory in index {
-                log::debug!("  Discovered advisory: {advisory:?}");
-                progress.set_message(
-                    advisory
-                        .url
-                        .path()
-                        .rsplit_once('/')
-                        .map(|(_, s)| s)
-                        .unwrap_or(advisory.url.as_str())
-                        .to_string()
-                        .into(),
-                );
-                visitor
-                    .visit_advisory(&context, advisory)
+
+            if let Some(directory_url) = &distribution.directory_url {
+                log::warn!("Walking: {:?}", directory_url.clone());
+                let index = self
+                    .source
+                    .load_index(DiscoverContext {
+                        discover_context_type: DiscoverContextType::DirectoryUrl,
+                        url: directory_url.clone(),
+                    })
                     .await
-                    .map_err(Error::Visitor)?;
-                progress.tick();
+                    .map_err(Error::Source)?;
+
+                let progress = self.progress.start(index.len());
+
+                for advisory in index {
+                    log::debug!("  Discovered advisory: {advisory:?}");
+                    progress.set_message(
+                        advisory
+                            .url
+                            .path()
+                            .rsplit_once('/')
+                            .map(|(_, s)| s)
+                            .unwrap_or(advisory.url.as_str())
+                            .to_string()
+                            .into(),
+                    );
+                    visitor
+                        .visit_advisory(&context, advisory)
+                        .await
+                        .map_err(Error::Visitor)?;
+                    progress.tick();
+                }
+            }
+            if let Some(rolie) = distribution.rolie {
+                for feed in rolie.feeds {
+                    log::warn!("Rolie feeds walking: {:?}", feed.url);
+                    let index = self
+                        .source
+                        .load_index(DiscoverContext {
+                            discover_context_type: DiscoverContextType::FeedUrl,
+                            url: feed.url.clone(),
+                        })
+                        .await
+                        .map_err(Error::Source)?;
+
+                    let progress = self.progress.start(index.len());
```

Comment:
Wouldn't it be possible to reduce the code duplication here? It look pretty similar to the previous `if` block.